Emscripten support take 2 (with bonus i686-unknown-linux-musl).

brson opened the pull request:
Here's another go at adding emscripten support. This needs to wait again on new [libc definitions](https://github.com/rust-lang-nursery/libc/pull/122) landing. To get the libc definitions right I had to add support for i686-unknown-linux-musl, which are very similar to emscripten's, which are derived from arm/musl.

This branch additionally removes the makefile dependency on the `EMSCRIPTEN` environment variable by not building the unused compiler-rt.

Again, this is not sufficient for actually compiling to asmjs since it needs additional LLVM patches.

r? @alexcrichton

aidanhs commented on src/librustc_trans/back/write.rs:
```diff
@@ -524,11 +529,21 @@ unsafe fn optimize_and_codegen(cgcx: &CodegenContext,
-        llvm::LLVMWriteBitcodeToFile(llmod, out.as_ptr());
+    // Change what we write and cleanup based on whether obj files are
+    // just llvm bitcode. In that case write bitcode, and possibly
+    // delete the bitcode if it wasn't requisted. Don't generate the
```
requested

aidanhs commented on src/libstd/sys/unix/thread.rs:
```diff
@@ -180,7 +182,7 @@ pub mod guard {
 
 
-#[cfg(any(target_os = "linux",
+#[cfg(any(all(target_os = "linux", not(target_env = "musl")),
```
Unfortunately I'm without the ability to build musl-rust right now, so I'm forced to ask rather than just trying - what about this won't work on musl?

brson replied:
I haven't looked too deeply, but I was getting crashes in one of the pthread calls in this code path, and since it's some hairy code I just punted.

alexcrichton commented on src/librustc_back/target/i686_unknown_linux_musl.rs:
```diff
@@ -0,0 +1,83 @@
+    // To deal with these circular dependencies we just force the compiler to
+    // link everything as a group, not stripping anything out until everything
+    // is processed. The linker will still perform a pass to strip out object
+    // files but it won't do so until all objects/archives have been processed.
```
Could the comments in this file be deduplicated with the 64-bit MUSL? e.g. just leave pointers saying that an explanation of all the weird options are over there.

alexcrichton commented:
r=me with one minor nit, but otherwise this'll just need libc to land first as well

bors commented:
:umbrella: The latest upstream changes (presumably #30601) made this pull request unmergeable. Please resolve the merge conflicts.

brson commented:
Rebasing this now.

brson commented:
Well, I've rebased and addressed comments, and this builds, but testing against emscripten incoming it no longer produces working js.

Still, I'm up for landing this for the wip.

brson commented:
@bors r=alexcrichton

bors commented:
:pushpin: Commit b9dd828 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton b9dd828c769a6e61a0f10e023be5f3ae25b4c990 -->

brson commented:
@bors r- I've got a commit that changes jemalloc in there.

brson commented:
cc https://github.com/kripken/emscripten/issues/4082

brson commented:
@bors r=alexcrichton

bors commented:
:pushpin: Commit 7afb56f has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 7afb56f51e52030fbdda07c38f9ae09a1daeed9f -->

bors commented:
:hourglass: Testing commit 7afb56f with merge 8c604dc...

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/7978), [auto-linux-32-opt](http://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/7999), [auto-linux-64-debug-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-debug-opt/builds/1246), [auto-linux-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/7939), [auto-linux-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/7949), [auto-linux-64-x-android-t](http://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/7939), [auto-linux-cross-opt](http://buildbot.rust-lang.org/builders/auto-linux-cross-opt/builds/1152), [auto-linux-musl-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-musl-64-opt/builds/3104), [auto-mac-32-opt](http://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/7985), [auto-mac-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-mac-64-nopt-t/builds/7987), [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/7952), [auto-mac-ios-opt](http://buildbot.rust-lang.org/builders/auto-mac-ios-opt/builds/122), [auto-win-gnu-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-nopt-t/builds/2990), [auto-win-gnu-32-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt/builds/3006), [auto-win-gnu-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-nopt-t/builds/2999), [auto-win-gnu-64-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-opt/builds/2981), [auto-win-msvc-32-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-32-opt/builds/2214), [auto-win-msvc-64-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt/builds/2984)

tamird commented:
@brson @alexcrichton a bit late to this party, but `base.pre_link_args.push("-Wl,-melf_i386".to_string());` is the only bit of this that isn't shared with the 64-bit version and it isn't clear why. Can you shed some light on this difference?

alexcrichton commented:
AFAIK that's just to tell the linker to emit a 32-bit executable in case it's default a 64-bit one

tamird commented:
Does that make it redundant with `-m32`?

On Mon, May 2, 2016 at 1:10 AM, Alex Crichton <notifications@github.com>
wrote:

> AFAIK that's just to tell the linker to emit a 32-bit executable in case
> it's default a 64-bit one
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/rust-lang/rust/pull/30629#issuecomment-216106038>
>

alexcrichton commented:
In this case, probably yes

brson commented:
@tamird Memory is hazy, but if I didn't write that it didn't work, and I don't know if it's redundant with `-m32`.
